MediaGalleryMetadata/Model/ExtractMetadata.php

```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryMetadata\Model;
+
+use Magento\MediaGalleryMetadataApi\Model\ReaderPool;
+use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterface;
+use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterfaceFactory;
+use Magento\MediaGalleryMetadataApi\Api\ExtractMetadataInterface;
+
+/**
+ * Extract metadata from the asset by path
+ */
+class ExtractMetadata implements ExtractMetadataInterface
+{
+    /**
+     * @var ReaderPool
+     */
+    private $readerPool;
+
+    /**
+     * @var MetadataInterfaceFactory
+     */
+    private $metadataFactory;
+
+    /**
+     * @param ReaderPool $readerPool
+     * @param MetadataInterfaceFactory $metadataFactory
+     */
+    public function __construct(ReaderPool $readerPool, MetadataInterfaceFactory $metadataFactory)
+    {
+        $this->readerPool = $readerPool;
+        $this->metadataFactory = $metadataFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(string $path): MetadataInterface
+    {
+        $title = '';
+        $description = '';
+        $keywords = [];
+        foreach ($this->readerPool->get() as $reader) {
+            $data = $reader->execute($path);
+            $title = $data->getTitle() ?? $title;
+            $description = $data->getDescription() ?? $description;
+            $keywords = array_unique(array_merge($keywords, $data->getKeywords()));
+            if (!empty($title) && !empty($description) && !empty($keywords)) {
+                break;
+            }
+        }
```

Comment:
I personally find this hard to understand, we are executing the reader from all formats on one file, each of those will return a MetadataInterface, after extracting the values and validating them we create a new MetadataInterface to return.

I would prefer to have a class that receives the path and returns the correct ExtractMetadataInterface for that file, also I think readerPool is misleading, I think the name extractMetadataPool would be more straight forward, I got confused and thought the readerPool was a pool of FileReader classes.

Reply:
Returning the correct ExtractMetadataInterface will:
1. Expose internal implementation
2. Make the client implementation more complex
3. Will not allow using multiple ExtractMetadataInterface implementation to process a single image from different angles)

Reply:
Aggree with `extractMetadataPool`. Btw, as for the naming, do you think we should use `extractMetadata` term or `getMetadata` term for all classes?

Reply:
I think `extractMetadata` makes sense, we are actually doing more than just getting it, I mean, usually gets are simple functions, and here we are a more complex functionality.